contributors/design-proposals/add-on/coredump-daemonset.md

```diff
@@ -0,0 +1,115 @@
+   * determine process in which pod/namespace generated the core file
+   * save coredump metadata into api-server
+   * save coredump file into persistent volume
+2. A coredump-controller which support quota of coredump files in each namesapce
+![architecture](./architecture.png)
+
+# core_pattern
+To determine whether a core file is generated for process in a k8s container, we
+override /proc/sys/kernel/core_pattern kernel parameter in kubelet node.
+```
+|/coredump/coredump-detector -P=%P -p=%p -e=%e -t=%t -c=/coredump/config --log_dir=/coredump/
+```
+
+When coredump happens, linux kernel will call coredump-detector and give core
+dump file as standard input to coredump-detector.
+coredump-detector will:
+* access the docker api and distinguish where(which container) the core dump comes from
+* access the kubernetes cluster and distinguish which namespace this pod belongs to
+* register the coredump metadata to api-server
+* save core dump file to local host cache
+
+# custom resource definition
+CustomResourceDefinition (CRD) is a built-in API of kubernetes that offers a simple way
+to create custom resources. We created [two CRDs](https://github.com/CaoShuFeng/coredump-detector/blob/master/yaml/coredump-crd.yaml) to save our own
+custom resource into the key-value storage, `coredumps` and `coredumpquotas`.
+`coredumps` stores the metadata of coredump files:
+```go
+type Coredump struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata"`
+        Spec              CoredumpSpec   `json:"spec"`
+        Status            CoredumpStatus `json:"status,omitempty"`
+}
+
+type CoredumpSpec struct {
+        ContainerName string    `json:"containerName"`
+        Pod           string    `json:"pod"`
+        Uid           types.UID `json:"uid"`
+        Pid           int       `json:"pid"`
+        Filename      string    `json:"filename"`
+        // Time is the kernel time when coredump happens.
+        Time metav1.Time `json:"dumptime"`
+        // Volume is the persistent volume, where coredump file is saved.
+        Volume string `json:"volume"`
+        // Size of coredump file
+        Size *resource.Quantity `json:"size"`
+}
+
+type CoredumpStatus struct {
+        State   CoredumpState `json:"state,omitempty"`
+        Message string        `json:"message,omitempty"`
+}
+```
+
+`coredumpquotas` defines the quota of coredump in each namespace:
+```go
+type CoredumpQuota struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata"`
+        Spec              QuotaSpec   `json:"spec"`
+        Status            QuotaStatus `json:"status"`
+}
+
+type QuotaSpec struct {
+        Hard *resource.Quantity `json:"hard"`
+}
+
+type QuotaStatus struct {
+        Used *resource.Quantity `json:"used"`
+        Hard *resource.Quantity `json:"hard"`
+}
+```
+
+# coredump-controller
+Now CRD in kubernetes doesn't support quota, so we deploy a controller to work as
```

Comment:
the plan is to support this soon once we can have a shared cache with garbage collection.

so in your case, we would just quota: `count/coredumps.coredump`

Reply:
Hi, I will update this according to the new feature.

Does quota for crd support resource.Quantity?
For example: we allow 2Gi coredump files in namespace A.
But not: we allow 200 coredump files in namespace A.